src/common/dav/dav_iface.c

```diff
@@ -57,43 +76,60 @@ setup_dav_phdr(dav_obj_t *hdl)
 static void
 persist_dav_phdr(dav_obj_t *hdl)
 {
-	pmemops_persist(&hdl->p_ops, hdl->do_phdr, offsetof(struct dav_phdr, dp_unused));
+	mo_wal_persist(&hdl->p_ops, hdl->do_phdr, offsetof(struct dav_phdr, dp_unused));
 }
 
 static dav_obj_t *
-dav_obj_open_internal(int fd, int flags, size_t sz)
+dav_obj_open_internal(int fd, int flags, size_t sz, const char *path)
 {
 	int rc;
 	dav_obj_t *hdl = NULL;
+	struct umem_wal_tx *utx = NULL;
 	void *base;
 	char *heap_base;
 	uint64_t heap_size;
 	int persist_hdr = 0;
+	int err = 0;
 
 	base = mmap(NULL, sz, PROT_READ|PROT_WRITE, MAP_SHARED, fd, 0);
 	if (base == MAP_FAILED) {
 		close(fd);
 		return NULL;
 	}
 
-	D_ALIGNED_ALLOC(hdl, 64, sizeof(dav_obj_t));
+	D_ALIGNED_ALLOC(hdl, CACHELINE_SIZE, sizeof(dav_obj_t));
 	if (hdl == NULL) {
-		errno = ENOMEM;
+		err = ENOMEM;
 		goto out1;
 	}
 	memset(hdl, 0, sizeof(dav_obj_t));
 
+	/* REVISIT: In future pass the meta instance as argument instead of fd */
+	hdl->do_mi = NULL;
 	hdl->do_fd = fd;
 	hdl->do_base = base;
 	hdl->do_size = sz;
+	hdl->p_ops.base = hdl;
+
+	D_ALLOC_PTR(utx);
+	if (utx == NULL) {
+		err = ENOMEM;
+		D_FREE(hdl);
+		goto out1;
+	}
+	hdl->do_utx = utx;
```

Comment:
The utx should be per transaction but not per pool. For md on SSD, when a TX yield on committing WAL, we should allow another TX start and run in parallel, otherwise, the IOPS will be way too low.

Reply:
The utx (or the wal txI) is not tied to the umem transaction. DAV transaction works very similar to pmemobj transaction. The only difference is pmemobj persists its changes using pmemobj_persit/flush calls and DAV does delayed persist by first pushing a corresponding redo log to utx and later the redo logs in utx is committed to wal at a safe point which is dav_tx_end(). The utx is re-allocated immediately after the commit. The utx gets updated at the commit stage of umem transaction or by umem atomics or by explicit calls to mo_wal_persist/flush. Adding an entry to utx implies that the entry has to committed in the next wal commit for the vos object. If a umem transaction fails there is no real need to commit all pending changes in utx or wal tx immediately. It can be committed by the next ULT doing an wal tx commit for the same vos file. Since the utx is part of dav, the next ULT that updates utx will inherit those pending changes and push it as part of the commit.